### DISCUSSION TITLE
Using ndk_http_module with nginx-unprivileged:1.25-alpine3.18

### DISCUSSION THREAD

**Discussion by irina-batmanova** [nginx/docker-nginx-unprivileged#187] [Q&A] [answered]
Hello,
I'm trying to use ndk_http_module (need this dependency for ngx_http_lua_module) with the latest nginx unpiveledged, seems like the latest version of this module that's available as an apk is 1.24 (checked [here](https://pkgs.alpinelinux.org/packages?name=nginx-mod-http-lua&branch=edge&repo=&arch=&maintainer=)), which is incomatible with nginx 1.25 (getting the following error message when trying to run nginx with this module installed:

> nginx: [emerg] module "/usr/lib/nginx/modules/ndk_http_module.so" version 1024000 instead of 1025003 in /etc/nginx/nginx.conf:1

Do I understand correctly that ndk_http_module 1.24 is incompatible with nginx 1.25, and if yes - is it often the case that nginx version is updated without updating ndk_http_module and ngx_http_lua_module? 
I'm switching from openresty to nginx + lua for the sake of getting updates to the base image more often, which would improve the security of my service. If I won't be able to update regularly because of the modules not being updated - switching from openresty won't make sense.

**Comment by alessfg** [comment, marked as answer]
Hey @irina-batmanova!  Let me first address some parts of your question, and then I'll give you a couple suggestions --

> ... seems like the latest version of this module that's available as an apk is 1.24 (checked [here](https://pkgs.alpinelinux.org/packages?name=nginx-mod-http-lua&branch=edge&repo=&arch=&maintainer=)) ... 

The link you shared lists the packages hosted by the official Alpine repo manager (apk). If you search for standalone NGINX, you'll see that the version available also happens to be 1.24. These are the latest "stable" packages, which are then compiled and hosted by the Alpine team. NGINX builds and hosts their own packages and maintains two development branches, the `stable` branch, and a `mainline` branch -- the latest mainline release, `1.25.3`, is the one contained in the image you are trying to use. (We recommend using the `mainline` branch for most if not all NGINX users.)

> ...and is it often the case that nginx version is updated without updating ndk_http_module and ngx_http_lua_module?...

All packages we build and host are updated to support the latest version of NGINX whenever a new version is released.

> Do I understand correctly that ndk_http_module 1.24 is incompatible with nginx 1.25?

Yes, you would need a matching release version.

 ---

All that being said, we sadly do not support either the NDK or Lua modules for NGINX Open Source (we only do for NGINX Plus). We do build those modules, but you would need to pull the packages from our mercurial repo directly (https://hg.nginx.org/pkg-oss/file/tip/contrib/src). It's not an ideal workflow, so a couple suggestions that come to mind are:
* If you can somehow make a normal/unrestricted Docker image work in your environment, I would suggest looking at the system in place in the core Docker NGINX repo that lets you build images with various 3rd party modules https://github.com/nginxinc/docker-nginx/tree/master/modules (with the caveat that those modules will **not** be supported). 
* A different alternative that **is** supported (yet will initially be more time consuming) and would work with these images is to look into porting your Lua code to NGINX JavaScript (NJS). It's a much more performant scripting language that's built **and** supported when using NGINX Open Source.

**Reply by irina-batmanova** [reply]
Thanks @alessfg! I think using NJS can be an option for us, as we're using just a small bit of Lua in our config.
